Add sample demonstrating VK_KHR_shader_relaxed_extended_instruction

gpx1000 opened the pull request:
## Description

Add a sample that shows how to use VK_KHR_shader_relaxed_extended_instruction and tutorial for it.

Fixes #<issue number>

## General Checklist:

Please ensure the following points are checked:

- [ ] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [ ] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [ ] I have commented any added functions (in line with Doxygen)
- [ ] I have commented any code that could be hard to understand
- [ ] My changes do not add any new compiler warnings
- [ ] My changes do not add any new validation layer errors or warnings
- [ ] I have used existing framework/helper functions where possible
- [ ] My changes do not add any regressions
- [ ] I have tested every sample to ensure everything runs correctly
- [ ] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [ ] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [ ] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

 If this PR contains framework changes:
 - [ ] I did a full batch run using the `batch` command line argument to make sure all samples still work properly

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [ ] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [ ] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [ ] Any dependent assets have been merged and published in downstream modules
- [ ] For new samples, I have added a paragraph with a summary to the appropriate chapter in the readme of the folder that the sample belongs to [e.g. api samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/api/README.adoc)
- [ ] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [ ] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

gpx1000 committed:
Add sample demonstrating VK_KHR_shader_relaxed_extended_instruction usage

gpx1000 committed:
fix copyright and clang-format

SaschaWillems commented on samples/extensions/shader_relaxed_extended_instruction/README.adoc:
```diff
@@ -0,0 +1,81 @@
+- See the License for the specific language governing permissions and
+- limitations under the License.
+-
+////
```
This is missing the "Where to get the source" Antora block (see readme.adoc for other samples or the readme from the samples template).

gpx1000 committed:
- Update README with GitHub repository link for sample

gary-sweet commented:
I'm not really sure what I'm expecting to see with this sample.

What I'm currently seeing is that I seem to have some seemingly random framebuffers begin displayed, and a bunch of these being printed to the console:
```
[info] vkQueueSubmit(): pSubmits[0] DebugPrintf:
relaxed-ext-inst demo: gid = 0
[info] vkQueueSubmit(): pSubmits[0] DebugPrintf:
relaxed-ext-inst demo: gid = 0
```

Keenuts commented:
Hello!
I looked at the SPIR-V file in this PR (`relaxed_demo.comp.spv`) and it seams it has no `OpExtInstWithForwardRef` instructions.

I don't know if GLSL emits those, but you can have a HLSL shader doing the same thing with the following code:

```hlsl
// RUN: %dxc %s -T cs_6_0 -spirv -fspv-target-env=vulkan1.3 -E main -fspv-debug=vulkan-with-source -O3

class A {
  void foo(uint3 gid) {
    printf("relaxed-ext-inst demo: gid = %u", gid.x);
  }
};

[numthreads(1, 1, 1)]
void main(uint3 gid : SV_DispatchThreadID)
{
  A a;
  a.foo(tid);
}
```

Here, one `OpExtInstWithForwardRefs` will be emitted as we have debug info with a class method, which brings a circular dependencies in the debug instructions.

gpx1000 committed:
Merge branch 'main' into VK_KHR_shader_relaxed_extended_instruction

# Conflicts:
#	samples/extensions/README.adoc

gpx1000 committed:
Convert shader_relaxed_extended_instruction sample from GLSL to Slang and update copyrights to 2026

gpx1000 commented:
Great thanks for the validation of HLSL.  I went ahead and switched from GLSL to Slang and validated that it does emit the OpExtInstWithForwardRef instruction.  This Sample should now be ready (also merged the latest master and fixed merge conflicts.)

gpx1000 committed:
Convert shader_relaxed_extended_instruction sample from GLSL to Slang with debug info to trigger OpExtInstWithForwardRef

gpx1000 committed:
Update copyrights to 2026

Keenuts commented on samples/extensions/shader_relaxed_extended_instruction/README.adoc:
```diff
@@ -0,0 +1,109 @@
+}
+----
+
+Shader (GLSL) used by this sample:
```
this is not up-to-date no?

Keenuts commented on shaders/shader_relaxed_extended_instruction/slang/relaxed_demo.comp.slang:
```diff
@@ -0,0 +1,47 @@
+// would emit OpExtInstWithForwardRef instructions when compiled with rich
+// debug info (e.g., NonSemantic.Shader.DebugInfo.100).
+//
+// Note: As of Slang in Vulkan SDK 1.4.328, the compiler does not yet emit
```
Why using slang if it cannot generate the instruction?
The DXC version provided in the same SDK supports it and will emit the proper instruction.
Looking at the disassembly of the associated `.spv` file, the extended instruction this sample is supposed to demonstrate is not generated.

You can use the HLSL sample I gave, which compiles and does generate the opcode you are showcasing in this sample.

gpx1000 replied:
Migrated to HLSL and using the exact shader you sent.  Sorry for that.

gary-sweet commented:
I'm still just seeing three random garbage frames cycling on the display. Will need some investigation.

asuessenbach commented:
On my platform `VK_EXT_layer_settings` is not available.
On startup, I get a couple of warnings:
```
[warning] 2132353751 - VALIDATION-SETTINGS: vkCreateInstance(): Application is using deprecated "enables" (VK_LAYER_ENABLES) layer settings.
Deprecated settings and new settings cannot be mixed, and deprecated ones take precedence. Consider only using the new settings:
  Deprecated: VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT | New: "printf_enable" (VK_LAYER_PRINTF_ENABLE=1)
  
[warning] 2132353751 - VALIDATION-SETTINGS: vkCreateInstance(): DebugPrintf logs to the Information message severity, enabling Information level logging otherwise the message will not be seen.  

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing fragmentStoresAndAtomics to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing vertexPipelineStoresAndAtomics to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing shaderInt64 to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceTimelineSemaphoreFeatures to pNext with timelineSemaphore set to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceVulkanMemoryModelFeatures to pNext with vulkanMemoryModel and vulkanMemoryModelDeviceScope set to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceBufferDeviceAddressFeatures to pNext with bufferDeviceAddress set to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceScalarBlockLayoutFeatures to pNext with scalarBlockLayout set to VK_TRUE

[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDevice8BitStorageFeatures to pNext with storageBuffer8BitAccess set to VK_TRUE
```
And then an endless list of
```
[info] vkQueueSubmit(): pSubmits[0] DebugPrintf:
relaxed-ext-inst demo: gid = 0
```

The actual sample windows stays black.

Is that to be expected?

gpx1000 committed:
Convert shader_relaxed_extended_instruction sample from Slang to HLSL/DXC with UI-driven dispatch and message capture

Replaces Slang shader with HLSL variant using DXC that reliably emits OpExtInstWithForwardRef patterns. Adds interactive UI with value slider and manual dispatch button, captures debugPrintf messages via debug utils callback, and eliminates per-frame dispatch to avoid console spam. Updates to Vulkan 1.3 target and pre-records UI render pass commands.

gpx1000 commented:
I implemented a GUI and allow user to send a number that will be printed out from the shader and routed back to the GUI.  I also am displaying in the GUI if the sample is supporting the extension being used.  As our framework already enables vkdebugprintf in shaders, you will see the print outs even if you don't have the extension supported.  However, in the case that the user were to implement this on a system that supports it, then they can use the relaxed extended instruction to not need to do extra work and see their shader print out.
In this design, there's nothing really interesting to display as I'm demonstrating using this extension to enable debug without enabling it, so I tried to give a GUI.. so it at least *looks* more interesting than a black screen.  If anyone has thoughts for what else I can do; lemme know.  But I've intentionally kept this sample simple.

gary-sweet commented on samples/extensions/shader_relaxed_extended_instruction/shader_relaxed_extended_instruction.cpp:
```diff
@@ -0,0 +1,366 @@
+{
+	if (drawer.header("VK_KHR_shader_relaxed_extended_instruction"))
+	{
+		const bool has_ext  = get_instance().is_enabled(VK_KHR_SHADER_RELAXED_EXTENDED_INSTRUCTION_EXTENSION_NAME);
```
get_instance().is_enabled doesn't appear to report device extensions. I see OFF for both extensions when they are both actually supported.

gary-sweet reviewed (commented):
I am seeing the debug messages appear in the UI, but only if I build in debug mode, is that expected? In release mode, no messages ever appear.

gpx1000 commented:
The messages in the UI will appear in the case that Validation Layers and debugprintf are enabled, that happens in the framework when we do a debug build.  So, artifact of that.
The fact that get_instance().is_enabled isn't working is a surprise.  I'll track that down in the morning.

gary-sweet commented:
> The messages in the UI will appear in the case that Validation Layers and debugprintf are enabled, that happens in the framework when we do a debug build. So, artifact of that. The fact that get_instance().is_enabled isn't working is a surprise. I'll track that down in the morning.

Should we make it obvious somewhere that this only works fully in debug mode? I can just see people (like me) thinking that it just doesn't work.

gpx1000 commented:
This one, technically is more technical.  See, the point of the extension is if you don't enable debugprintf for the shader in your code, you can still use it as it's a relaxed extended instruction.  Thus, the real bug is on your system that does support the extension you're getting it behaving like the extension doesn't exist.  In Release and Debug with the extension you'll get it behaving properly.  Without the extension's help, you'll get it behaving "properly" due to us making that a convenient call.

gpx1000 commented:
actually found the bug...  I'm an idiot.  device extension not instance extension.

gpx1000 committed:
Fix extension check to query device extensions instead of instance extensions

The VK_KHR_shader_relaxed_extended_instruction and VK_KHR_shader_non_semantic_info are device extensions, not instance extensions. Update the check to use `get_device().is_extension_enabled()` instead of `get_instance().is_enabled()`.

gpx1000 committed:
remove frustration comment :)

Keenuts reviewed (commented):
Shader in the latest revision LGTM (demonstrated opcodes are present) Thanks!

SaschaWillems commented:
Does compile and run for me (Win 11, RTX 4070), but I do get several warnings.

First these:

```
[warning] 2132353751 - VALIDATION-SETTINGS: vkCreateInstance(): Application is using deprecated "enables" (VK_LAYER_ENABLES) layer settings.
Deprecated settings and new settings cannot be mixed, and deprecated ones take precedence. Consider only using the new settings:
  Deprecated: VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT | New: "printf_enable" (VK_LAYER_PRINTF_ENABLE=1)

[warning] 2132353751 - VALIDATION-SETTINGS: vkCreateInstance(): DebugPrintf logs to the Information message severity, enabling Information level logging otherwise the message will not be seen.
[info] Found GPU: NVIDIA GeForce RTX 4070
```

And at a later point these:

```
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing fragmentStoresAndAtomics to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing vertexPipelineStoresAndAtomics to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Forcing shaderInt64 to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceTimelineSemaphoreFeatures to pNext with timelineSemaphore set to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceVulkanMemoryModelFeatures to pNext with vulkanMemoryModel and vulkanMemoryModelDeviceScope set to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceBufferDeviceAddressFeatures to pNext with bufferDeviceAddress set to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDeviceScalarBlockLayoutFeatures to pNext with scalarBlockLayout set to VK_TRUE
[warning] -2030147807 - WARNING-Setting-Limit-Adjusted: vkCreateDevice(): Internal Warning: Adding a VkPhysicalDevice8BitStorageFeatures to pNext with storageBuffer8BitAccess set to VK_TRUE
```

gpx1000 committed:
Improve debug printf setup and handle missing validation layer gracefully

Switch to boolean `printf_enable` layer setting (removing string-based `enables`), fix indentation, and make validation layer optional in fallback path by checking availability before adding to enabled layers list.

gpx1000 committed:
Fix indentation from spaces to tabs in shader_relaxed_extended_instruction sample Loading

asuessenbach commented on samples/extensions/shader_relaxed_extended_instruction/shader_relaxed_extended_instruction.cpp:
```diff
@@ -210,58 +239,40 @@ bool ShaderRelaxedExtendedInstruction::prepare(const vkb::ApplicationOptions &op
-	VK_CHECK(vkEndCommandBuffer(cmd));
-}
+void ShaderRelaxedExtendedInstruction::record_minimal_present_cmd(VkCommandBuffer) const
+{}
```
Empty function?

asuessenbach reviewed (commented):
Just one minor issue.

And I get the very same warnings as Sascha.

Besides that: looks good.

gpx1000 committed:
Merge branch 'main' into VK_KHR_shader_relaxed_extended_instruction

SaschaWillems commented:
Sample works fine.

One question though, is this still true as of today?

> At the time of writing, Slang and glslc in the Vulkan SDK does not emit the necessary debug information patterns that lead to OpExtInstWithForwardRef in SPIR-V for this demonstration. The HLSL/DXC path above does emit the required SPIR-V relaxed extended instruction sequence, so this sample uses HLSL for reliability.

Keenuts commented:
> One question though, is this still true as of today?
> 
> > At the time of writing, Slang and glslc in the Vulkan SDK does not emit the necessary debug information patterns that lead to OpExtInstWithForwardRef in SPIR-V for this demonstration. The HLSL/DXC path above does emit the required SPIR-V relaxed extended instruction sequence, so this sample uses HLSL for reliability.|

Just tried on slang 2025.23.2 (LunarG latest SDK release 1.4.335.0) and it does not emit the correct opcode. Don't know about mainline or next SDK version to be released soon)

gpx1000 committed:
Merge branch 'main' into VK_KHR_shader_relaxed_extended_instruction

# Conflicts:
#	framework/vulkan_type_mapping.h
#	samples/extensions/README.adoc

gpx1000 committed:
Refactor debugPrintf setup to use framework hooks with manual fallback demonstration

Replaces manual instance creation with framework request hooks (`request_layer_settings`, `request_layers`, etc.) for modern VK_EXT_layer_settings path. Retains explicit VK_EXT_validation_features fallback in `create_instance()` override for educational purposes when layer settings unavailable. Adds missing `<algorithm>` and `<cstring>` includes, returns VK_API_VERSION_1_3 from `get_api_version()`, and improves logging clarity.

gpx1000 committed:
fix clang-format.
